### PR TITLE
Simpy compatibility layer

### PR DESCRIPTION
**Pull request by maxfischer2781**
Allow running SimPy simulations inside uSim. This pull request includes:

* the ``usim.py`` package, a drop-in replacement for ``simpy``,
* basic SimPy functionality of ``Environment`` and ``Event``s,
* the ability to run ``simpy`` standalone or embedded into ``usim``,
* the ability for uSim activities to ``await`` SimPy events,
* documentation including migration advice.

This pull request does *not* include:

* emulation of ``simpy.util`` and ``simpy.resources`` (to be added by another PR),
* the ability for SimPy processes to yield uSim notifications (to be added by another PR),
* emulation of ``simpy.rt`` real-time support.

---
This change is [Reviewable](https://reviewable.io/reviews/mainekuehn/usim/25)

### CODE REVIEW

File: usim/py/events.py [hunk 1]
```diff
@@ -0,0 +1,642 @@
...
+        except TypeError:
+            raise AttributeError(
+                f"{self.__class__.__name__!r} object has no attribute 'value' set yet"
+            )
+        else:
+            return value if exception is None else exception
+
+    def trigger(self, event: 'Event') -> 'Event':
+        """
+        Trigger this event to match the state of ``event``
```
Comment (eileen-kuehn) [inline usim/py/events.py]: What do you mean with *match the state of event*?

File: usim/py/events.py [hunk 2]
```diff
@@ -0,0 +1,642 @@
...
+class Initialize:
+    """
+    Internal initializer for :py:class:`~.Process`
+
+    Since this event is internal and not used by 'usim.py',
+    it is not implemented. Instantiating this class
+    unconditionally raises an error.
+    """
+    def __init__(self, *args, **kwargs):
+        raise NotEmulatedError(self.__doc__)
```
Comment (eileen-kuehn) [inline usim/py/events.py]: Should be adapted.

File: usim/py/events.py [hunk 3]
```diff
@@ -0,0 +1,642 @@
...
+class Process(Event[V]):
+    """
+    A concurrently running process which triggers as an event on completion
+
+    Processes wrap around generators, which in turn may ``yield`` events to
+    wait for them. This implements cooperative multitasking, allowing another
+    process to run while others wait. Each :py:class:`~.Process` acts as an
+    event which is automatically triggered when the underlying generator completes.
+
+    .. code:: python3
```
Comment (eileen-kuehn) [inline usim/py/events.py]: I am afraid the examples don't speak for themselves. What do you want to show? E.g. what is the purpose of `clocks`? Is it only to show the difference to the `multi_clock` function below? Then I wouldn't get the point when reading until here.

File: usim/py/events.py [hunk 4]
```diff
@@ -0,0 +1,642 @@
...
+                    await (time + 1)
+
+        To run an activity, one must distinguish between a concurrent
+        :py:class:`~usim.typing.Task` and a nested activity. When an activity
+        exclusively waits for another activity, it can ``await`` it.
+
+        .. code:: python3
+
+            async def double_clock(sound):
+                "A clock of double duration"
```
Comment (eileen-kuehn) [inline usim/py/events.py]: This cannot be copied to try out the example. You could use a common comment here, for example.

File: usim/py/events.py [hunk 5]
```diff
@@ -0,0 +1,642 @@
...
+
+        When migrating from :py:class:`~.Process`, activities are expressed as
+        ``async def`` coroutines instead of generators. Activities do not need
+        an environment - it is implicitly available. An activity can ``await``
+        a :term:`notification`, concurrently iterate using ``async for``
+        and manage resources using ``async with``.
+
+        .. code:: python3
+
+            async def clock(sound):
```
Comment (eileen-kuehn) [inline usim/py/events.py]: Is it on purpose to have two clock functions? The example below uses this clock function and not the other above. So by just copying the code users might run into problems.

File: usim/py/events.py [hunk 6]
```diff
@@ -0,0 +1,642 @@
...
+                await clock(sound)
+                await clock(sound)
+
+        Only concurrent activities must be handled as a :py:class:`~usim.typing.Task`.
+        Use a :py:class:`usim.Scope` to open new tasks and manage their lifetime:
+
+        .. code:: python3
+
+            async def multi_clock():
+                "A clock of making multiple sounds"
```
Comment (eileen-kuehn) [inline usim/py/events.py]:
```suggestion
                "A clock making multiple sounds"
```

Further you should again adapt the comment.

File: usim/py/events.py [hunk 7]
```diff
@@ -0,0 +1,642 @@
...
+            except (StopIteration, StopProcess) as err:
+                value = err.args[0] if err.args else None
+                self.succeed(value)
+                break
+            except BaseException as err:
+                self.fail(err)
+                break
+
+    @property
+    def is_alive(self):
```
Comment (eileen-kuehn) [inline usim/py/events.py]: This method might be better suitable in the `Event` itself.

File: usim/py/events.py [hunk 8]
```diff
@@ -0,0 +1,642 @@
...
+    def __repr__(self):
+        return (
+            f'<usim.py.{self.__class__.__name__} object'
+            f' of {self._generator.__name__!r} at {id(self)}>'
+        )
+
+
+class ConditionValue:
+    """
+    ``dict``-like view to the events used in a condition and their value
```
Comment (eileen-kuehn) [inline usim/py/events.py]: Do they all share one value or do they have a value each?

```suggestion
    ``dict``-like view to the events used in a condition and their values
```

File: usim/py/events.py [hunk 9]
```diff
@@ -0,0 +1,642 @@
...
+
+    .. note::
+
+        This type only captures the events, not their values.
+        If the value of an event changes, this class reflects the change.
```
Comment (eileen-kuehn) [inline usim/py/events.py]: I think the class itself cannot reflect the change but the instance of the class. How correct do you want to be?

File: usim/py/events.py [hunk 10]
```diff
@@ -0,0 +1,642 @@
...
+
+    On any change of events, ``evaluate`` is called with two arguments:
+    ``events`` is a tuple of the events passed to the condition,
+    and ``num_triggered`` is the number of events that have triggered so far.
+    This can be used to quickly test how many events have triggered:
+
+    .. code:: python3
+
+        def most(events, count):
+            "Test that at least 50% of events have triggered"
```
Comment (eileen-kuehn) [inline usim/py/events.py]: Again, the comment issue.

File: usim/py/events.py [hunk 11]
```diff
@@ -0,0 +1,642 @@
...
+            for event in unobserved[:]:
+                if not event.__usimpy_flag__:
+                    continue
+                elif event.ok:
+                    unobserved.remove(event)
+                    observed.append(event)
+                else:
+                    event.defused = True
+                    self.fail(event.value)
+                    return
```
Comment (eileen-kuehn) [inline usim/py/events.py]: For this branch you could have a small function as you are using it in two places, on the top for observed and here.

File: usim/py/events.py [hunk 12]
```diff
@@ -0,0 +1,642 @@
...
+                self.fail(event.value)
+                return
+        while unobserved and not self._evaluate(self._events, len(observed)):
+            await AnyFlag(*(event.__usimpy_flag__ for event in unobserved))
+            for event in unobserved[:]:
+                if not event.__usimpy_flag__:
+                    continue
+                elif event.ok:
+                    unobserved.remove(event)
+                    observed.append(event)
```
Comment (eileen-kuehn) [inline usim/py/events.py]: From this point on you could think about only having a count for `observed` events as you are not accessing the events again. So you could introduce a local variable, initialise before while with `len(observed)` and then only do the increment.

File: usim/py/events.py [hunk 13]
```diff
@@ -0,0 +1,642 @@
...
+            self.succeed(ConditionValue(*self._flatten_values(self._events)))
+
+    @classmethod
+    def _flatten_values(cls, events) -> List[Event]:
+        """Flatten the values of all events"""
+        result = []
+        for event in events:
+            if isinstance(event, Condition):
+                result.extend(cls._flatten_values(event._events))
+            elif event.ok:
```
Comment (eileen-kuehn) [inline usim/py/events.py]: Why is here the distinction between `Condition` and `event.ok`, what about `event.fail`? (I suppose a failure is also a kind of value)

File: usim/py/exceptions.py [hunk 14]
```diff
@@ -0,0 +1,40 @@
+"""
+This module defines the exceptions exclusive to the SimPy compatibility layer.
+These are used as shutdown signals between the environment and processes.
+Note that with the exception of :py:exc:`~usim.py.exceptions.Interrupt`,
+all signals should be considered internal and not meant to be handled manually.
```
Comment (eileen-kuehn) [inline usim/py/exceptions.py]:
```suggestion
all signals are considered internal and are not meant to be handled manually.
```

File: usim/py/exceptions.py [hunk 15]
```diff
@@ -0,0 +1,40 @@
...
+"""
+
+
+class NotEmulatedError(NotImplementedError):
```
Comment (eileen-kuehn) [inline usim/py/exceptions.py]: This Error we should consider for renaming. As mentioned in several other comments, the compatibility layer does not do a proper emulation but a simulation. However, talking about simulation error inside a simulation might be confusing for a user. So we should think about a proper name.

File: usim/py/exceptions.py [hunk 16]
```diff
@@ -0,0 +1,40 @@
...
+    """An operation of the 'simpy' API is not emulated by the 'usim.py' API"""
```
Comment (eileen-kuehn) [inline usim/py/exceptions.py]:
```suggestion
    """An operation of the 'simpy' API is not mapped by the 'usim.py' API"""
```

File: usim/py/exceptions.py [hunk 17]
```diff
@@ -0,0 +1,40 @@
...
+
+
+class StopSimulation(BaseException):
+    """Signal to stop a simulation"""
+
+
+class SimPyException(Exception):
+    """Base case for all exceptions of ``usim.py`` that can safely be handled"""
```
Comment (eileen-kuehn) [inline usim/py/exceptions.py]: What happens to exceptions from SimPy that cannot be handled safely?

File: usim/py/exceptions.py [hunk 18]
```diff
@@ -0,0 +1,40 @@
...
+
+
+class StopProcess(SimPyException):
+    """
+    Signal to stop a process
+
+    .. warning::
+
+        This exceptions exists for historical compatibility only.
```
Comment (eileen-kuehn) [inline usim/py/exceptions.py]:
```suggestion
        This exception exists for historical compatibility only.
```

File: usim_pytest/test_usimpy/test_core.py [hunk 19]
```diff
@@ -0,0 +1,79 @@
...
+        output = []
+        env.process(observable(env, output))
+        env.run()
+        assert output == [1, 2]
+        assert env.now == 2
+
+    def test_until_past(self, env):
+        """env.run(until) argument must not be in the past"""
+        with pytest.raises(ValueError):
+            env.run(-5)
```
Comment (eileen-kuehn) [inline usim_pytest/test_usimpy/test_core.py]: Consider also testing edge cases, such as `0`. I think I put a comment somewhere, that `0` should not raise a `ValueError`.

File: docs/source/topics/simpy.rst [hunk 20]
```diff
@@ -0,0 +1,103 @@
...
+    --- The SimPy Documentation
+
+Both μSim and SimPy are discrete, event-based simulation frameworks.
+Whereas SimPy is built on generators and callbacks, μSim relies exclusively
+on the ``async``/``await`` coroutine support introduced in Python 3.5.
+This makes a difference both in the usage, as well as supported features.
+
+To allow running, integrating and migrating existing SimPy simulations,
+μSim has an inbuilt compatibility layer: the :py:mod:`usim.py` module
+emulates most of the SimPy API. The only truly unsupported feature are scheduling
```
[the remark this answers is not in the document]
Comment (maxfischer2781) [reply docs/source/topics/simpy.rst]: We could go with one of "imitate" or "replicate". What do you think?

File: docs/source/topics/simpy.rst [hunk 21]
```diff
@@ -0,0 +1,103 @@
...
+    ...     while True:
+    ...         print('Start parking at %d' % env.now)
+    ...         parking_duration = 5
+    ...         yield env.timeout(parking_duration)
+    ...
+    ...         print('Start driving at %d' % env.now)
+    ...         trip_duration = 2
+    ...         yield env.timeout(trip_duration)
+    ...
+    >>> import usim.py as simpy  # import usim.py instead of simpy
```
[the remark this answers is not in the document]
Comment (maxfischer2781) [reply docs/source/topics/simpy.rst]: The SimPy example consists of two separate code snippets - the second one starts at the import. So it replicates the example, but is not exakt either way.

We could split this back up into two parts, showing that the first doesn't need any modification and the second one just the changed import. Sounds good?

File: usim/py/core.py [hunk 22]
```diff
@@ -0,0 +1,319 @@
...
+The latter allows interactions between the μSim and SimPy components.
+"""
+from typing import Optional, List, Tuple, Coroutine, Generator, TypeVar, Iterable,\
+    Union
+from .._core.loop import __LOOP_STATE__, Loop, ActivityError
+from .. import time, run as usim_run
+from .. import Scope
+
+from .events import Event
+from .exceptions import NotEmulatedError, StopSimulation, StopProcess
```
[the remark this answers is not in the document]
Comment (maxfischer2781) [reply usim/py/core.py]: This is related to not saying that we emulate simpy?

File: usim/py/core.py [hunk 23]
```diff
@@ -0,0 +1,319 @@
...
+            else:
+                if isinstance(until, Event):
+                    if until.processed:
+                        return until.value
+                    raise RuntimeError("'until' event was not triggered")
+        else:
+            raise NotEmulatedError(
+                "'env.run' is not supported inside a 'usim' simulation\n"
+                "\n"
+                "Synchronous running blocks the event loop. Use instead\n"
```
[the remark this answers is not in the document]
Comment (maxfischer2781) [reply usim/py/core.py]:
```suggestion
                "Synchronous 'run' block the event loop. Use instead:\n"
```
Matching wording of similar error messages.

File: usim/py/core.py [hunk 24]
```diff
@@ -0,0 +1,319 @@
...
+                "* 'await env.until()' to block only the current activity\n"
+                "* 'async with env:' to concurrently run the environment\n"
+                "\n"
+                "You can 'env.run' outside of a 'usim' simulation"
```
[the remark this answers is not in the document]
Comment (maxfischer2781) [reply usim/py/core.py]: How about "may"?

File: usim/py/core.py [hunk 25]
```diff
@@ -0,0 +1,319 @@
...
+
+    def exit(self, value=None):
+        """
+        Stop the current process, optionally providing a ``value``
+
+        .. warning::
+
+            This method exists for historical compatibility only.
+            Use ``return value`` instead.
+        """
```
[the remark this answers is not in the document]
Comment (maxfischer2781) [reply usim/py/core.py]: Older Python versions did not allow ``return`` in a generator. So SimPy processes could not ``return`` a value either. This really only exists for completeness. Should this be made clearer? I guess it *is* confusing without in-depth knowledge.

File: usim/py/core.py [hunk 26]
```diff
@@ -0,0 +1,319 @@
...
+        raise StopProcess(value)
+
+    def step(self):
+        """
+        'Environment.step' is not implemented by the μSim compatibility layer
+
+        The μSim compatibility layer uses the regular μSim event loop.
+        There is no public alternative to 'Environment.step'.
+        """
+        raise NotEmulatedError(self.step.__doc__)
```
[the remark this answers is not in the document]
Comment (maxfischer2781) [reply usim/py/core.py]: Any suggestion how? ;)

File: docs/source/topics/simpy.rst (hunk 20 above)
[the remark this answers is not in the document]
Comment (eileen-kuehn) [reply docs/source/topics/simpy.rst]: In this case we might even go with *map* as I proposed in one of the other docstrings.

File: usim/py/events.py [hunk 27]
```diff
@@ -0,0 +1,642 @@
...
+        #: List of callbacks to run when the event is triggered
+        self.callbacks = []  # type: List[Callable[[E], None]]
+        self._value = None  # type: Optional[Tuple[V, Optional[BaseException]]]
+        #: Whether a failure of this event has been handled
+        self.defused = False
+
+    # Implementation Note
+    # ``usim`` would flatten Conditions of the same type here. With the
+    # ``simpy`` model, this does not work well. The problem is that events
+    # are active components that may fail the Environment.
```
[the remark this answers is not in the document]
Comment (maxfischer2781) [reply usim/py/events.py]: I am willing to be less-than exact in these comments - they are hopefully only relevant for people that murk around in the internals anyways. In this case, it should make sense at least in hindsight.

File: usim/py/events.py (hunk 1 above)
Comment (maxfischer2781) [reply usim/py/events.py]: In SimPy, ``self`` copies the result and several flags from ``event``. In ``usim.py``, this is all folded into the result. How about this?

```suggestion
        Trigger this event and inherit the value or exception of ``event``
```

File: usim/py/events.py (hunk 7 above)
Comment (maxfischer2781) [reply usim/py/events.py]: Agreed, but the SimPy API is like this. I can implement it via the corresponding Event method, though.

File: usim/py/events.py (hunk 4 above)
Comment (maxfischer2781) [reply usim/py/events.py]:
```suggestion
                '''A clock of double duration'''
```

File: usim/py/events.py (hunk 9 above)
Comment (maxfischer2781) [reply usim/py/events.py]: How about this?

```suggestion
        If the value of a contained event changes, the value in the ConditionValue changes as well.
```

File: usim/py/events.py (hunk 10 above)
Comment (maxfischer2781) [reply usim/py/events.py]:
```suggestion
            '''Test that at least 50% of events have triggered'''
```

File: usim/py/events.py (hunk 13 above)
Comment (maxfischer2781) [reply usim/py/events.py]: For a Condition, we *must* recurse to flatten it - this is because ``|`` and ``&`` produce a tree, not a chain. (See above comment why my attempts to fix this failed miserably.) If any event is ``event.fail``, then the entire Condition has failed already.

I guess an ``assert`` may be appropriate, both for documentation and robustness.

File: usim/py/exceptions.py (hunk 17 above)
Comment (maxfischer2781) [reply usim/py/exceptions.py]: The ``StopSimulation`` (and ``NotEmulatedError ``) is not supposed to be handled manually. How bout this?
```suggestion
    """Base case for all exceptions of ``usim.py`` that can safely be handled in a simulation"""
```

File: docs/source/topics/simpy.rst (hunk 21 above)
[the remark this answers is not in the document]
Comment (maxfischer2781) [reply docs/source/topics/simpy.rst]: I've split this example now. Makes the change much clearer.

File: usim/py/core.py [hunk 28]
```diff
@@ -0,0 +1,319 @@
...
+        return await self._scope.__aexit__(exc_type, exc_val, exc_tb)
+
+    async def until(self, until=None):
+        """Asynchronous version of the :py:meth:`~.run` method"""
+        async with self:
+            if until is not None:
+                if isinstance(until, Event):
+                    await until.__usimpy_flag__
+                else:
+                    if until <= time.now:
```
[the remark this answers is not in the document]
Comment (maxfischer2781) [reply usim/py/core.py]: Allowed ``until == time.now`` and added unit test.

**Comment by maxfischer2781** [comment]
I've either directly committed or implemented most things addressed. We should decide on a better word for "emulate" and try and come up with an example for ``Process``.

File: usim/py/events.py (hunk 7 above)
Comment (maxfischer2781) [reply usim/py/events.py]: Turns out ``Event`` doesn't have anything like this. The closes expression would be ``not self.triggered``, but that seems equally unhelpful. I recommend keeping it like this, any objections?

File: usim/py/events.py (hunk 13 above)
Comment (maxfischer2781) [reply usim/py/events.py]: So I tried to come up with *some* paper assertion here but I did not find one. The publicAPI of ``Event`` is surprisingly bare. What could work is this:

```python3
assert event._value is None or event._value[1] is None or event.defused
```
Not sure if that is helpful at this point.

File: usim/py/exceptions.py (hunk 17 above)
Comment (eileen-kuehn) [reply usim/py/exceptions.py]: You could also say *during a simulation*. Change as you prefer :)

**Comment by maxfischer2781** [comment]
Renamed ``NotEmulatedError`` to a more specific ``NotCompatibleError`` for things we *cannot* (reasonably) provide -- things we just did not do are a regular ``NotImplementedError``. Where appropriate, replaced "emulate" with alternatives such as "provide" or "recreate".

**Comment by maxfischer2781** [comment]
I've swapped the Process example from clocks to travelling.

File: usim/py/events.py [hunk 29]
```diff
@@ -0,0 +1,658 @@
...
+        To run an activity, one must distinguish between a concurrent
+        :py:class:`~usim.typing.Task` and a nested activity. When an activity
+        exclusively waits for another activity, it can directly ``await`` it.
+
+        .. code:: python3
+
+            async def visit_many(*destinations: str):
+                '''Visit all destinations one after the other'''
+                for destination in destinations:
+                    # awat a bare activity to wait for its completion
```
Comment (eileen-kuehn) [inline usim/py/events.py]:
```suggestion
                    # await a bare activity to wait for its completion
```

File: usim/py/__init__.py [hunk 30]
```diff
@@ -0,0 +1,126 @@
...
+
+    μSim simulations can ``await`` events and may
+    :py:meth:`~usim.py.events.Event.succeed` and :py:meth:`~usim.py.events.Event.fail`
+    them.
+
+
+Resources
+---------
+
+``usim.py`` implements no resources yet.
```
[the remark this answers is not in the document]
Comment (maxfischer2781) [reply usim/py/__init__.py]: Not the ``simpy`` resources. Since they are not needed to get the compatibility layer running, I would prefer adding them in a later commit.

File: usim/py/core.py [hunk 31]
```diff
@@ -0,0 +1,321 @@
...
+        return False
+    else:
+        return True
+
+
+V = TypeVar('V')
+
+
+class Environment:
+    """
```
Comment (maxfischer2781) [inline usim/py/core.py]: Needed to allow ``\ `` escape.

```suggestion
    r"""
```
